### DISCUSSION TITLE
Built-in Lets Encrypt failing to issue cert

### DISCUSSION THREAD

**Discussion by coatmaker618** [requarks/wiki#3427] [Help / Questions]
**Question**

TL;DR:
Trying to get HTTPS cert from built-in Lets Encrypt but I get an error instead :(

Full story:
I am trying to utilize the built-in feature of lets-encrypt to newer versions of Wiki.js for issuing an https cert.

The domain works correctly for HTTP, so I believe the issue is isolated to the certbot/wikijs side of things.  However I am not sure how to debug this more beyond the output below:

Host Info:
OS: Ubuntu 18.04
Wiki.js version: 2.4.107
Database engine: mysql  (note: postgres is officially recommended over mysql.  That question is here: #2287  once this is fixed, I will happily update this to say postgres!)

output from letsencrypt portion of wikijs init (from running "node server"):
```
2020-08-11T14:17:47.201Z [MASTER] info: (LETSENCRYPT) Initializing Let's Encrypt client...
2020-08-11T14:17:47.255Z [MASTER] info: HTTP Server: [ RUNNING ]
2020-08-11T14:17:47.463Z [MASTER] info: Telemetry is active: [ OK ]
2020-08-11T14:17:47.755Z [MASTER] info: (LETSENCRYPT) Generating server keypairs...
2020-08-11T14:17:48.034Z [MASTER] info: (LETSENCRYPT) Server keypairs generated successfully [ OK ]
2020-08-11T14:17:48.035Z [MASTER] info: (LETSENCRYPT) Generating certificate signing request (CSR)...
2020-08-11T14:17:48.049Z [MASTER] info: (LETSENCRYPT) CSR generated successfully [ OK ]
2020-08-11T14:17:48.050Z [MASTER] info: (LETSENCRYPT) Requesting certificate from Let's Encrypt...
2020-08-11T14:17:48.060Z [MASTER] info: (LETSENCRYPT) Setting HTTP challenge for wiki.mink.ninja: [ READY ]
2020-08-11T14:17:48.061Z [MASTER] info: (LETSENCRYPT) Waiting for challenge to complete...
2020-08-11T14:17:53.325Z [MASTER] info: (LETSENCRYPT) Removing HTTP challenge: [ OK ]
2020-08-11T14:17:53.326Z [MASTER] info: (LETSENCRYPT) Removing HTTP challenge: [ OK ]
2020-08-11T14:17:53.327Z [MASTER] warn: (LETSENCRYPT) Error: unable to verify the first certificate
2020-08-11T14:17:53.327Z [MASTER] error: unable to verify the first certificate
```

**Comment by Smankusors** [comment]
Btw, is your Wiki.js behind reverse proxy?

**Comment by coatmaker618** [comment]
My wiki.js is behind a HAProxy acting as a reverse proxy.  However that HAProxy is ONLY doing reverse proxy and is not touching the TLS/cert side of things (which I know is NOT the default behavior for such configs).  So HTTPS is still coming through unmodified.

**Comment by Smankusors** [comment]
Hmm then the HTTPS should be handled by HAProxy, not by Wiki.js

**Comment by coatmaker618** [comment]
Interesting! Why is that?  I've seen that mentioned elsewhere but I thought that was only if the proxy was breaking https (which would totally make sense) & converting the traffic to HTTP.

This strikes me as signifigantly less secure as the traffic is unencrypted from the proxy to wikijs server.

**Comment by Smankusors** [comment]
> Interesting! Why is that? I've seen that mentioned elsewhere but I thought that was only if the proxy was breaking https (which would totally make sense) & converting the traffic to HTTP.
> 
> This strikes me as signifigantly less secure as the traffic is unencrypted from the proxy to wikijs server.

isn't your proxy and wiki.js are in the same internal network? Then it's ok too have it unencrypted IMO. Unless both of them exposed to public then it's different story.

**Comment by coatmaker618** [comment]
I see what you're saying, but it feels like adding unnecessary risk to me.  Why create a vulnerability, even a small one, when it can (seemingly) be easily avoided?  If there's a technical reason that this is preferred, please let me know!

After all, Lets Encrypt is "built-in" to wikijs to the point that you just change a few lines in the config & it (should) automatically handle the certs :)

**Comment by Smankusors** [comment]
> I see what you're saying, but it feels like adding unnecessary risk to me. Why create a vulnerability, even a small one, when it can (seemingly) be easily avoided? If there's a technical reason that this is preferred, please let me know!

why wasting a lot of time, if someone outside cannot sniff/MITM the HTTP on your network? It's only you that can do it, nobody else. If anyone can, then you sure have misconfiguration on your network, or your hosts are compromised. 😉

Btw, it's not easily executed, it's not simple as that, more on that below.

> After all, Lets Encrypt is "built-in" to wikijs to the point that you just change a few lines in the config & it (should) automatically handle the certs :)

LE designed to create certificates for most directly public facing servers. But behind reverse proxy? Well yes but no.

It's kinda difficult to execute. After the verification succeed, you need to change address to it's configured host name on reverse proxy. And in your reverse proxy you need to somehow trick the DNS by using `/etc/hosts` or something else. Or your reverse proxy might got invalid certificate warning or even infinite loop. 😵

Welp no, LE is not designed for SSL behind reverse proxy. Also you are not intended to grab SSL certificate that generated by Wiki.js. It's better to generate your own self signed certificate for this behind reverse proxy scenario.

---

Disclaimer : I haven't tried this, it's just theory on my mind.

**Comment by coatmaker618** [comment]
When you're saying:

> LE designed to create certificates for most directly public facing servers. But behind reverse proxy? Well yes but no.

> It's kinda difficult to execute. After the verification succeed, you need to change address to it's configured host name on reverse proxy. And in your reverse proxy you need to somehow trick the DNS by using /etc/hosts or something else. Or your reverse proxy might got invalid certificate warning or even infinite loop. 😵

I think we're talking about two different things.  Because my reverse proxy is ONLY redirecting the subdomains to different machines.  It's not changing ports or addresses or hostnames of anything! (well ok, in a case like this is could change ports where wikijs wants to be on port 3000...but even that would just be a blind redirect from 80 to 3000).


Also when you say
> why wasting a lot of time, if someone outside cannot sniff/MITM the HTTP on your network? It's only you that can do it, nobody else. If anyone can, then you sure have misconfiguration on your network, or your hosts are compromised. 😉

Then I say that you have far more trust in your ability to setup a network properly than I do in me 😛   I have made my share of stupid mistakes (and I'm sure I'm far from done), and would prefer to limit the damage when I do so.

Also worth pointing out that I have done this with certbot on several other systems in this same setup.  The instructions they provide work exactly as described.  No jiggery-pokery required!

**Comment by coatmaker618** [comment]
I think it may actually be more accurate to say that, while I do have HAProxy running between the web & my wikijs...it may not, technically, be acting as a reverse proxy in the sense that nothing appears to be coming "from" HAProxy (user side or wikijs side).

I apologize for the confusion, but they do call it reverse proxy when you need to route traffic based on subdomain, even if that server can still "see"the internet.

**Comment by Smankusors** [comment]
The reverse proxy is not as simple as redirecting. It's creating another HTTP packet, an entirely different packet. That's why you will use different SSL certificate for that.  When there's HTTPS packet coming, the reverse proxy will decrypt the packet, and then create a new another packet to Wiki.js. If you enable SSL on Wiki.js side, the reverse proxy will also encrypt that new packet.

> Then I say that you have far more trust in your ability to setup a network properly than I do in me 😛 I have made my share of stupid mistakes (and I'm sure I'm far from done), and would prefer to limit the damage when I do so.

IMO, you need these stupid mistakes to be a better person. Without failure experience, you never know 😉

**Comment by coatmaker618** [comment]
> The reverse proxy is not as simple as redirecting. It's creating another HTTP packet, an entirely different packet. That's why you will use different SSL certificate for that. When there's HTTPS packet coming, the reverse proxy will decrypt the packet, and then create a new another packet to Wiki.js. If you enable SSL on Wiki.js side, the reverse proxy will also encrypt that new packet.

This is why I'm realizing it's better to say my setup is NOT doing Reverse Proxying.  My HAProxy does not have any certs at all on it, and I have several services running Certbot/LE without an issue.  I was using the wrong terminology because they call it "Reverse Proxy" when you set it up.

>IMO, you need these stupid mistakes to be a better person. Without failure experience, you never know 😉

Totally agree, this is why I assume I will make more and try to limit the damage when I do!!


Also I updated previous comments before I saw your response, sorry if they came in after you started typing your response.

I'm still looking forward to getting LE working through wikijs :)

**Comment by coatmaker618** [comment]
So having establish that my wikijs install is indeed NOT behind a reverse proxy--any advice for setting up the cert?

**Comment by Smankusors** [comment]
> So having establish that my wikijs install is indeed NOT behind a reverse proxy--any advice for setting up the cert?

if you still want to use built in let's encrypt functions, you can read more about it [here](https://docs.requarks.io/install/config#lets-encrypt)

**Comment by coatmaker618** [comment]
Right, I've tried putting those parameters in my setup & I got the error above :(

Is there any port forwarding that needs to be done for LE?

**Comment by Smankusors** [comment]
> Right, I've tried putting those parameters in my setup & I got the error above :(
> 
> Is there any port forwarding that needs to be done for LE?

hmm.... What port forwardings did you do? Did you change HTTPS port when port forwarding?

**Comment by coatmaker618** [comment]
I redirected port 80 to 3000 & 443 to 3443 as per the wiki to work

**Comment by Smankusors** [comment]
Welp I think the different port confuses Wiki.js... Perhaps you can port forward by 80 <-> 80 and 443 <-> 443?

**Comment by coatmaker618** [comment]
I certainly could do that, but doesn't wiki.js need the traffic to go to port 3000?   Or could I change that to port 80?

**Comment by Smankusors** [comment]
> I certainly could do that, but doesn't wiki.js need the traffic to go to port 3000? Or could I change that to port 80?

well it's mentioned on the [same page](https://docs.requarks.io/install/config#http) I previously mentioned to you 😉

**Comment by coatmaker618** [comment]
*facepalm* So it is!   I had it in my head that wiki.js couldn't run on port 80 or something, for some reason.....  Whelp, I'll just switch that config file to port 80/443 and hopefully be up and running with Lets Encrypt soon enough  (by which I mean onto issue #2287)

**Comment by coatmaker618** [comment]
Damn, I was so certain switching to port 80/443 was going to work!

Interestingly, I don't see anything regarding (LETS ENCRYPT) which I showed above.


```
Loading configuration from /home/server/wiki-v001/config.yml... OK
2020-08-18T02:00:58.303Z [MASTER] info: =======================================
2020-08-18T02:00:58.308Z [MASTER] info: = Wiki.js 2.4.107 =====================
2020-08-18T02:00:58.309Z [MASTER] info: =======================================
2020-08-18T02:00:58.310Z [MASTER] info: Initializing...
2020-08-18T02:00:59.128Z [MASTER] info: Using database driver pg for postgres [ OK ]
2020-08-18T02:00:59.158Z [MASTER] info: Connecting to database...
2020-08-18T02:00:59.199Z [MASTER] info: Database Connection Successful [ OK ]
2020-08-18T02:00:59.268Z [MASTER] warn: DB Configuration is empty or incomplete. Switching to Setup mode...
2020-08-18T02:00:59.268Z [MASTER] info: Starting setup wizard...
2020-08-18T02:00:59.468Z [MASTER] info: Starting HTTP server on port 80...
2020-08-18T02:00:59.469Z [MASTER] info: HTTP Server on port: [ 80 ]
2020-08-18T02:00:59.474Z [MASTER] info: HTTP Server: [ RUNNING ]
2020-08-18T02:00:59.474Z [MASTER] info: 🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻
2020-08-18T02:00:59.475Z [MASTER] info:
2020-08-18T02:00:59.476Z [MASTER] info: Browse to http://localhost:80/ to complete setup!
2020-08-18T02:00:59.477Z [MASTER] info:
2020-08-18T02:00:59.477Z [MASTER] info: 🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺
2020-08-18T02:01:00.387Z [MASTER] error: Not Found
TypeError: Cannot read property 'sendError' of undefined
```

**Comment by Smankusors** [comment]
well I think you need to finish the setup first... but got that unknown error instead...

maybe... you got corrupted installation? Can you try redownload it maybe?

**Comment by coatmaker618** [comment]
> well I think you need to finish the setup first... but got that unknown error instead...

Agreed!  Now if only my install would let me :p

As for downloading, I thought git did SHA as part of pull, so there should be no way I got a bad version.  But hey, git pull is quick enough to run anyway so sure.

Will update when I try with a pull into a fresh local repo.

**Comment by coatmaker618** [comment]
Just ran `wget https://github.com/Requarks/wiki/releases/download/2.4.107/wiki-js.tar.gz` to download again.  I got the same result (and same sha256 hash) :(

which, for comparison is:
`ac5c389015578d145f440763ad68724bdf652691c7334fef5fe4b5c695cdc033`

**Comment by coatmaker618** [comment]
I was able to change my setup to have nodejs listen on port 80, I think?

since it's not easily available, it's ``sudo setcap 'cap_net_bind_service=+ep' /usr/bin/node`` on Ubuntu

However, now I see no mention of LETSENCRYPT in the initial load phase (which seems bad):


```
2020-08-22T02:20:31.384Z [MASTER] info: =======================================
2020-08-22T02:20:31.389Z [MASTER] info: = Wiki.js 2.4.107 =====================
2020-08-22T02:20:31.389Z [MASTER] info: =======================================
2020-08-22T02:20:31.390Z [MASTER] info: Initializing...
2020-08-22T02:20:32.307Z [MASTER] info: Using database driver pg for postgres [ OK ]
2020-08-22T02:20:32.335Z [MASTER] info: Connecting to database...
2020-08-22T02:20:32.376Z [MASTER] info: Database Connection Successful [ OK ]
2020-08-22T02:20:32.446Z [MASTER] warn: DB Configuration is empty or incomplete. Switching to Setup mode...
2020-08-22T02:20:32.447Z [MASTER] info: Starting setup wizard...
2020-08-22T02:20:32.644Z [MASTER] info: Starting HTTP server on port 80...
2020-08-22T02:20:32.645Z [MASTER] info: HTTP Server on port: [ 80 ]
2020-08-22T02:20:32.649Z [MASTER] info: HTTP Server: [ RUNNING ]
2020-08-22T02:20:32.650Z [MASTER] info: 🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻🔻
2020-08-22T02:20:32.650Z [MASTER] info:
2020-08-22T02:20:32.651Z [MASTER] info: Browse to http://localhost:80/ to complete setup!
2020-08-22T02:20:32.651Z [MASTER] info:
2020-08-22T02:20:32.652Z [MASTER] info: 🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺🔺
```
